tests/test_halohash.py

```diff
@@ -0,0 +1,206 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# purldb is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/purldb for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+from collections import defaultdict
+
+import csv
+import math
+import copy
+import os
+import random
+import subprocess
+
+from commoncode.testcase import FileBasedTesting
+
+from matchcode_toolkit import halohash
+
+
+SIZE_IN_BITS = 128
+
+
+def load_csv(location):
+    """
+    Load a CSV file at location and return a tuple of (field names, list of rows as
+    mappings field->value).
+    """
+    with open(location) as csvin:
+        reader = csv.DictReader(
+            csvin,
+            quoting=csv.QUOTE_NONNUMERIC
+        )
+        fields = reader.fieldnames
+        values = sorted(reader, key=lambda d: d.items())
+        return fields, values
+
+
+def check_results(
+    results,
+    expected_file,
+    fieldnames,
+    regen=False,
+):
+    """
+    Load and compare the CSV at `expected_file` against `results`.
+    """
+    if regen:
+        with open(expected_file, 'w') as f:
+            writer = csv.DictWriter(
+                f,
+                fieldnames=fieldnames,
+                quoting=csv.QUOTE_NONNUMERIC
+            )
+            writer.writeheader()
+            writer.writerows(results)
+
+    field_names, expected = load_csv(expected_file)
+    key1 = field_names[0]
+    # then check results line by line for more compact results
+    for exp, res in zip(expected, results):
+        assert exp[key1] == res[key1]
+        expected_mean_hamming_distance = exp['mean hamming distance']
+        exp_min = expected_mean_hamming_distance - exp['standard deviation']
+        exp_max = expected_mean_hamming_distance + exp['standard deviation']
+        assert exp_min <= res['mean hamming distance']
```

Comment:
@pombredanne I'm having trouble getting tests to pass, where the mean hamming distance value of our test run is larger or smaller than the expected maximum or minimum hamming distance value.

```
            ...
            assert exp_min <= res['mean hamming distance']
>           assert res['mean hamming distance'] <= exp_max
E           assert 9.4 <= 9.299999999999999
```

Do you have any suggestions on how to best compare results?

Reply:
I would try to bump the acceptable range to be +/-20%

Reply:
@pombredanne The same issue is happening:
```
            expected_mean_hamming_distance = exp['mean hamming distance']
            exp_min = expected_mean_hamming_distance * 0.8
            exp_max = expected_mean_hamming_distance * 1.2
            assert exp_min <= res['mean hamming distance']
>           assert res['mean hamming distance'] <= exp_max
E           assert 6.2 <= 5.88
```

Reply:
Increasing number of tests per word removed/replaced gives us a better standard deviation.